File: AssetManager/DataComms/Munis/Munis_Comms.vb
﻿Imports System.Data.SqlClient

Public Class Munis_Comms

#Region "Fields"

    Private Const MSSQLConnectString As String = "server=svr-munis5.core.co.fairfield.oh.us; database=mu_live; trusted_connection=True;"

#End Region

#Region "Methods"

    Public Function Return_MSSQLCommand(strSQLQry As String) As SqlCommand
        Dim conn As SqlConnection = New SqlConnection(MSSQLConnectString)
        Dim cmd As New SqlCommand
        cmd.Connection = conn
        cmd.CommandText = strSQLQry
        Return cmd
    End Function

    Public Function Return_MSSQLTable(strSQLQry As String) As DataTable
        Using conn As SqlConnection = New SqlConnection(MSSQLConnectString),
                    NewTable As New DataTable,
                    da As New SqlDataAdapter
            da.SelectCommand = New SqlCommand(strSQLQry)
            da.SelectCommand.Connection = conn
            da.Fill(NewTable)
            Return NewTable
        End Using
    End Function

    Public Async Function Return_MSSQLTableAsync(strSQLQry As String) As Task(Of DataTable)
        Using conn As SqlConnection = New SqlConnection(MSSQLConnectString),
                    NewTable As New DataTable,
                cmd As New SqlCommand(strSQLQry, conn)
            Await conn.OpenAsync
            Dim dr As SqlDataReader = Await cmd.ExecuteReaderAsync
            NewTable.Load(dr)
            Return NewTable
        End Using
    End Function

    Public Function Return_MSSQLTableFromCmd(cmd As SqlCommand) As DataTable
        Using NewTable As New DataTable,
                    da As New SqlDataAdapter(cmd)
            da.Fill(NewTable)
            cmd.Dispose()
            Return NewTable
        End Using
    End Function

    Public Async Function Return_MSSQLTableFromCmdAsync(cmd As SqlCommand) As Task(Of DataTable)
        Using conn = cmd.Connection,
                NewTable As New DataTable
            Await conn.OpenAsync
            Dim dr As SqlDataReader = Await cmd.ExecuteReaderAsync
            NewTable.Load(dr)
            cmd.Dispose()
            Return NewTable
        End Using
    End Function


#End Region

End Class

File: AssetManager/Data/DataComms/Munis/MunisComms.cs
using Databases.Data;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;
using System.Threading.Tasks;

namespace AssetManager.Data.Communications
{
    public static class MunisComms
    {
        private const string msSqlConnectString = "server=svr-munis5.core.co.fairfield.oh.us; database=mu_live; trusted_connection=True; Connection Timeout=5";

        public static SqlCommand ReturnSqlCommand(string query)
        {
            SqlConnection conn = new SqlConnection(msSqlConnectString);
            SqlCommand cmd = new SqlCommand();
            cmd.Connection = conn;
            cmd.CommandText = query;
            return cmd;
        }

        public static DataTable ReturnSqlTable(string query)
        {
            using (SqlConnection conn = new SqlConnection(msSqlConnectString))
            using (DataTable newTable = new DataTable())
            using (SqlDataAdapter da = new SqlDataAdapter())
            {
                da.SelectCommand = new SqlCommand(query);
                da.SelectCommand.Connection = conn;
                da.Fill(newTable);
                return newTable;
            }
        }

        public static async Task<DataTable> ReturnSqlTableAsync(string query)
        {
            using (SqlConnection conn = new SqlConnection(msSqlConnectString))
            using (DataTable newTable = new DataTable())
            using (SqlCommand cmd = new SqlCommand(query, conn))
            {
                await conn.OpenAsync();
                SqlDataReader dr = await cmd.ExecuteReaderAsync();
                newTable.Load(dr);
                return newTable;
            }
        }

        public static DataTable ReturnSqlTableFromCmd(SqlCommand cmd)
        {
            using (DataTable newTable = new DataTable())
            using (SqlDataAdapter da = new SqlDataAdapter(cmd))
            {
                da.Fill(newTable);
                cmd.Dispose();
                return newTable;
            }
        }

        public static async Task<DataTable> ReturnSqlTableFromCmdAsync(SqlCommand cmd)
        {
            using (var conn = cmd.Connection)
            using (DataTable newTable = new DataTable())
            {
                await conn.OpenAsync();
                SqlDataReader dr = await cmd.ExecuteReaderAsync();
                newTable.Load(dr);
                cmd.Dispose();
                return newTable;
            }
        }

        public static object ReturnSqlValue(string table, object fieldIn, object valueIn, string fieldOut, object fieldIn2 = null, object valueIn2 = null)
        {
            string query = "";
            QueryParamCollection queryParams = new QueryParamCollection();

            if (fieldIn2 != null && valueIn2 != null)
            {
                query = "SELECT TOP 1 " + fieldOut + " FROM " + table;

                queryParams.Add(fieldIn.ToString(), valueIn.ToString(), true);
                queryParams.Add(fieldIn2.ToString(), valueIn2.ToString(), true);
            }
            else
            {
                query = "SELECT TOP 1 " + fieldOut + " FROM " + table;

                queryParams.Add(fieldIn.ToString(), valueIn.ToString(), true);
            }
            using (var cmd = GetSqlCommandFromParams(query, queryParams.Parameters))
            using (var conn = cmd.Connection)
            {
                cmd.Connection.Open();
                return cmd.ExecuteScalar();
            }
        }

        public static async Task<string> ReturnSqlValueAsync(string table, object fieldIn, object valueIn, string fieldOut, object fieldIn2 = null, object valueIn2 = null)
        {
            string query = "";

            QueryParamCollection queryParams = new QueryParamCollection();

            if (fieldIn2 != null && valueIn2 != null)
            {
                query = "SELECT TOP 1 " + fieldOut + " FROM " + table;

                queryParams.Add(fieldIn.ToString(), valueIn.ToString(), true);
                queryParams.Add(fieldIn2.ToString(), valueIn2.ToString(), true);
            }
            else
            {
                query = "SELECT TOP 1 " + fieldOut + " FROM " + table;

                queryParams.Add(fieldIn.ToString(), valueIn.ToString(), true);
            }

            using (var cmd = GetSqlCommandFromParams(query, queryParams.Parameters))
            using (var conn = cmd.Connection)
            {
                await cmd.Connection.OpenAsync();
                var Value = await cmd.ExecuteScalarAsync();
                if (Value != null)
                {
                    return Value.ToString();
                }
            }
            return string.Empty;
        }

        /// <summary>
        /// Takes a partial query string without the WHERE operator, and a list of <see cref="DBQueryParameter"/> and returns a parameterized <see cref="SqlCommand"/>.
        /// </summary>
        /// <param name="partialQuery"></param>
        /// <param name="parameters"></param>
        /// <returns></returns>
        public static SqlCommand GetSqlCommandFromParams(string partialQuery, List<DBQueryParameter> parameters)
        {
            var cmd = ReturnSqlCommand(partialQuery);
            cmd.CommandText += " WHERE";
            string paramString = "";
            int valueSeq = 1;
            foreach (var fld in parameters)
            {
                if (fld.IsExact)
                {
                    paramString += " " + fld.FieldName + "=@Value" + valueSeq.ToString();
                    cmd.Parameters.AddWithValue("@Value" + valueSeq.ToString(), fld.Value);
                }
                else
                {
                    paramString += " " + fld.FieldName + " LIKE CONCAT('%', @Value" + valueSeq.ToString() + ", '%')";
                    cmd.Parameters.AddWithValue("@Value" + valueSeq.ToString(), fld.Value);
                }

                // Add operator if we are not on the last parameter.
                if (parameters.IndexOf(fld) < parameters.Count - 1)
                {
                    paramString += " " + fld.OperatorString;
                }

                valueSeq++;
            }

            cmd.CommandText += paramString;
            return cmd;
        }
    }
}